p2p/net/conngater/conngater.go

```diff
@@ -0,0 +1,322 @@
+	if ds != nil {
+		cg.ds = namespace.Wrap(ds, datastore.NewKey(ns))
+		cg.loadRules()
+	}
+
+	return cg
+}
+
+func (cg *BasicConnectionGater) loadRules() {
+	// load blocked peers
+	res, err := cg.ds.Query(query.Query{Prefix: keyPeer})
+	if err != nil {
+		log.Errorf("error querying datastore for blocked peers: %s", err)
+		return
+	}
+
+	for r := range res.Next() {
+		if r.Error != nil {
+			log.Errorf("query result error: %s", r.Error)
+			return
+		}
+
+		p := peer.ID(r.Entry.Value)
+		cg.blockedPeers[p] = struct{}{}
+	}
+
+	// load blocked addrs
+	res, err = cg.ds.Query(query.Query{Prefix: keyAddr})
+	if err != nil {
+		log.Errorf("error querying datastore for blocked addrs: %s", err)
+		return
+	}
+
+	for r := range res.Next() {
+		if r.Error != nil {
+			log.Errorf("query result error: %s", r.Error)
+			return
+		}
+
+		ip := net.IP(r.Entry.Value)
+		cg.blockedAddrs[ip.String()] = struct{}{}
+	}
+
+	// load blocked subnets
+	res, err = cg.ds.Query(query.Query{Prefix: keySubnet})
+	if err != nil {
+		log.Errorf("error querying datastore for blocked subnets: %s", err)
+		return
+	}
+
+	for r := range res.Next() {
+		if r.Error != nil {
+			log.Errorf("query result error: %s", r.Error)
+			return
+		}
+
+		ipnetStr := string(r.Entry.Value)
+		_, ipnet, err := net.ParseCIDR(ipnetStr)
+		if err != nil {
+			log.Errorf("error parsing CIDR subnet: %s", err)
+			return
+		}
+		cg.blockedSubnets[ipnetStr] = ipnet
+	}
+}
+
+// BlockPeer adds a peer to the set of blocked peers
+func (cg *BasicConnectionGater) BlockPeer(p peer.ID) {
+	cg.Lock()
+	defer cg.Unlock()
+
+	cg.blockedPeers[p] = struct{}{}
+
+	if cg.ds != nil {
+		err := cg.ds.Put(datastore.NewKey(keyPeer+p.String()), []byte(p))
+		if err != nil {
+			log.Errorf("error writing blocked peer to datastore: %s", err)
+		}
+	}
+}
+
+// UnblockPeer removes a peer from the set of blocked peers
+func (cg *BasicConnectionGater) UnblockPeer(p peer.ID) {
+	cg.Lock()
+	defer cg.Unlock()
+
+	delete(cg.blockedPeers, p)
+
+	if cg.ds != nil {
+		err := cg.ds.Delete(datastore.NewKey(keyPeer + p.String()))
+		if err != nil {
+			log.Errorf("error deleting blocked peer from datastore: %s", err)
+		}
+	}
+}
+
+// ListBlockedPeers return a list of blocked peers
+func (cg *BasicConnectionGater) ListBlockedPeers() []peer.ID {
+	cg.Lock()
+	defer cg.Unlock()
+
+	result := make([]peer.ID, 0, len(cg.blockedPeers))
+	for p := range cg.blockedPeers {
+		result = append(result, p)
+	}
+
+	return result
+}
+
+// BlockAddr adds an IP address to the set of blocked addresses
+func (cg *BasicConnectionGater) BlockAddr(ip net.IP) {
+	cg.Lock()
+	defer cg.Unlock()
+
+	cg.blockedAddrs[ip.String()] = struct{}{}
+
+	if cg.ds != nil {
+		err := cg.ds.Put(datastore.NewKey(keyAddr+ip.String()), []byte(ip))
+		if err != nil {
+			log.Errorf("error writing blocked addr to datastore: %s", err)
+		}
+	}
+}
+
+// UnblockAddr removes an IP address from the set of blocked addresses
+func (cg *BasicConnectionGater) UnblockAddr(ip net.IP) {
```

Comment:
Let's make all the block/unblock methods return an error. All of this needs to be reliable. It's a security-related component. If the user wants to simply log the error, they can. But let _them_ decide.